src/casync-http.c

```diff
@@ -11,67 +12,1066 @@
+                        r = -EIO;
+                        goto fail;
+                }
+        }
+
+        c = curl_easy_setopt(handle, CURLOPT_WRITEFUNCTION, write_callback);
+        if (c != CURLE_OK) {
+                log_error_curle(c, "Failed to set CURL callback function");
+                r = -EIO;
+                goto fail;
+        }
+
+        c = curl_easy_setopt(handle, CURLOPT_WRITEDATA, write_data);
+        if (c != CURLE_OK) {
+                log_error_curle(c, "Failed to set CURL callback data");
+                r = -EIO;
+                goto fail;
+        }
+
+        if (private) {
+                c = curl_easy_setopt(handle, CURLOPT_PRIVATE, private);
+                if (c != CURLE_OK) {
+                        log_error_curle(c, "Failed to set CURL private data");
+                        r = -EIO;
+                        goto fail;
+                }
+        }
+
+        /* (void) curl_easy_setopt(handle, CURLOPT_SSL_VERIFYPEER, false); */
+
+        /* (void) curl_easy_setopt(handle, CURLOPT_VERBOSE, 1L); */
+
+        *ret = handle;
+        return 0;
+
+fail:
+        curl_easy_cleanup(handle);
+        return r;
+}
+
+static int make_curl_multi_handle(CURLM **ret) {
+        int r;
+        CURLMcode c;
+        CURLM *handle = NULL;
+
+        assert(ret);
+        assert(*ret == NULL);
+
+        handle = curl_multi_init();
+        if (!handle) {
+                r = log_oom();
+                goto fail;
+        }
+
+        c = curl_multi_setopt(handle, CURLMOPT_MAX_HOST_CONNECTIONS, arg_max_host_connections);
+        if (c < 0) {
+                log_error_curlm(c, "Failed to set CURL max host connections");
+                r = -EIO;
+                goto fail;
+        }
+
+        /* Default since libcurl 7.62.0 */
+        c = curl_multi_setopt(handle, CURLMOPT_PIPELINING, CURLPIPE_MULTIPLEX);
+        if (c < 0)
+                log_error_curlm(c, "Failed to set CURL pipelining to multiplex, ignoring");
+
+        *ret = handle;
+        return 0;
+
+fail:
+        curl_multi_cleanup(handle);
+        return r;
+}
+
+/*
+ * chunks data
+ */
+
+typedef struct ChunkData ChunkData;
+
+struct ChunkData {
+        size_t current_store;  /* set to SIZE_MAX if chunk is missing */
+        CaChunkID id;
+        ReallocBuffer buffer;
+};
+
+static void chunk_data_reset(ChunkData *cd, CaChunkID *id) {
+        assert(cd);
+
+        cd->id = *id;
+        realloc_buffer_empty(&cd->buffer);
+}
+
+static void chunk_data_free(ChunkData *cd) {
+        if (cd == NULL)
+                return;
+
+        realloc_buffer_free(&cd->buffer);
+        free(cd);
+}
+
+static ChunkData *chunk_data_new(void) {
+        ChunkData *cd = NULL;
+
+        cd = new0(ChunkData, 1);
+
+        return cd;
+}
+
+/*
+ * simple queue implementation
+ */
+
+typedef struct QueueItem QueueItem;
+
+struct QueueItem {
+        void *data;
+        QueueItem *next;
+};
+
+typedef struct Queue {
```

Comment:
hmm, instead of introducing a new, local queue implementation here: let's just copy list.h from systemd's tree (i.e. https://github.com/systemd/systemd/blob/master/src/basic/list.h) and use that? it's an embedded list which means we need fewer allocations

Reply:
Done in commit 48f37ddf969b4bd4f7f0752c91df8df9575cf0ff and fdd46892bef4c8fd2774ccc82c45426913d60797. See my temporary branch [mr/casync-http-curl-multi](https://github.com/gportay/casync/commits/mr/casync-http-curl-multi).

I will push-force latter.